### ISSUE TITLE
Retrieving static resource via ServletContext root creates (many) jar_cache tmp files

### ISSUE THREAD

**Issue by BastiaanProot** [spring-projects/spring-boot#9866]
Since version 1.4.5 we notice our application produces jar_cache*.tmp files. In previous versions it did not. In our case the jar_cache*.tmp files risk filling the disk after a while.

We found the cause:

1. Locations for [static content](http://docs.spring.io/spring-boot/docs/current/reference/htmlsingle/#boot-features-spring-mvc-static-content) are [auto configured](https://github.com/spring-projects/spring-boot/blob/v1.5.4.RELEASE/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/web/ResourceProperties.java) by Spring Boot.
Default locations for static resources in the resolveChain (in order) are: 
	- (ServletContext) /
	- classpath:/META-INF/resources/
	- classpath:/resources/
	- classpath:/static/
	- classpath:/public/
1. In version 1.4.4 getting a resource from ServletContext root returned instances of *EmptyResource* and *getUrl()* returned *null*. So the **ServletContext location got skipped** and then resources were retrieved from _classpath:/META-INF/resources/_. **This changed in version 1.4.5: https://github.com/spring-projects/spring-boot/issues/8299**. ServletContext now returns a valid URL to the resource. 
1. Our static resources are contained in a jar in our application jar. So jar-in-jar. (URL to the resource looks like this: jar:war:file:application.jar*/BOOT_INF/lib/resources.jar!/META-INF/resources/favicon.ico) Retrieving the static resource requires JVM to use jar_cache*.tmp files. (using *File.deleteOnExit*). This seems to occur when you do [resource.contentLength()](https://github.com/spring-projects/spring-framework/blob/master/spring-webmvc/src/main/java/org/springframework/web/servlet/resource/ResourceHttpRequestHandler.java#L543).
On Debian/OpenJDK we can only see these jar_cache*.tmp files using **lsof**; they are open but deleted files. These do not get released until the application exits. So they still take up space and fill the disk. Similarly on Windows/SunJDK we can see the jar_cache*.tmp files accumulate and they are only deleted when the application exits. This seems to be a known/old phenomenon with the JVM.

In short:

Due to [changes in the behaviour of ServletContext static resources](https://github.com/spring-projects/spring-boot/issues/8299) our application now performs jar-in-jar lookups and accumulates tmp files.

Our solution:

We used the _spring.resources.static-locations_ application property to customize the order of the locations and put _classpath:/META-INF/resources/_ first.

**Comment by wilkinsona** [comment]
@BastiaanProot Thank you very much for the detailed analysis. We'd already seen the symptom you've described in #9751 but were some distance away from identifying the cause.

I see that you've mentioned `EmptyResource` so I assume that you're using Tomcat. Can you please confirm that's the case?

**Comment by BastiaanProot** [comment]
Indeed, we're using embedded Tomcat.

**Comment by wilkinsona** [comment]
The default ordering has been changed in https://github.com/spring-projects/spring-boot/commit/9dd3fb70e290bd426cd8377036f7174f9d8ddf3d that closed #9240.
